packages/protocol/contracts/common/MetaTransactionWallet.sol

```diff
@@ -0,0 +1,205 @@
+    eip712DomainSeparator = keccak256(
+      abi.encode(
+        keccak256(
+          "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+        ),
+        keccak256(bytes("MetaTransactionWallet")),
+        keccak256("1.1"),
+        id,
+        address(this)
+      )
+    );
+    emit EIP712DomainSeparatorSet(eip712DomainSeparator);
+  }
+
+  // For debugging purposes.
+  function getMetaTransactionDigest(
+    address destination,
+    uint256 value,
+    bytes memory data,
+    uint256 _nonce
+  ) public view returns (bytes32) {
+    bytes32 structHash = keccak256(
+      abi.encode(
+        EIP712_EXECUTE_META_TRANSACTION_TYPEHASH,
+        destination,
+        value,
+        keccak256(data),
+        _nonce
+      )
+    );
+    return keccak256(abi.encodePacked("\x19\x01", eip712DomainSeparator, structHash));
+  }
+
+  /**
+   * @notice Returns the address that signed the provided meta-transaction.
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param _nonce The nonce for this meta-transaction local to this wallet.
+   * @param v The recovery id of the ECDSA signature of the meta-transaction.
+   * @param r Output value r of the ECDSA signature.
+   * @param s Output value s of the ECDSA signature.
+   * @return The address that signed the provided meta-transaction.
+   */
+  function getMetaTransactionSigner(
+    address destination,
+    uint256 value,
+    bytes memory data,
+    uint256 _nonce,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) public view returns (address) {
+    bytes32 digest = getMetaTransactionDigest(destination, value, data, _nonce);
+    // TODO: Should we link, or inline, this library?
+    // Currently modified it to be inlined, but that will cause bytecode changes
+    // to all other contracts that use getSignerOfMessageHash.
+    return Signatures.getSignerOfMessageHash(digest, v, r, s);
+  }
+
+  /**
+   * @notice Executes a meta-transaction on behalf of the signer.`
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param v The recovery id of the ECDSA signature of the meta-transaction.
+   * @param r Output value r of the ECDSA signature.
+   * @param s Output value s of the ECDSA signature.
+   * @return The return value of the meta-transaction execution.
+   */
+  function executeMetaTransaction(
+    address destination,
+    uint256 value,
+    bytes calldata data,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    address _signer = getMetaTransactionSigner(destination, value, data, nonce, v, r, s);
+    require(_signer == signer, "Invalid meta-transaction signer");
```

Comment:
I'm pretty confused how the signing works, it looks like we get a signer but don't use it.

Reply:
The signer is the address that is authorized to call `executeTransaction` or sign meta transactions that can be executed by anyone via `executeMetaTransaction`

Reply:
Essentially you can think of it as the wallet contract's owner. The actual `owner` is the contract itself, so that `onlyOwner` functions can be executed by the signer via `executeTransaction` or `executeMetaTransaction`